### DISCUSSION TITLE
Using SnapStart for .NET8 throws InvalidStateTransition

### DISCUSSION THREAD

**Discussion by tmitchel2** [aws/aws-lambda-dotnet#1879] [Q&A] [answered]
I've updated a lambda function to use .NET8 runtime (instead of a custom container) to take advantage of the new SnapStart feature.  My function is invoked via ApiGateway.  The function works perfectly fine with all these changes and SnapStart disabled.  However, when I enable SnapStart the invocation always fails with the following:

```
2024-11-20T15:03:02.885Z		fail	Failed to invoke before snapshot hooks: HTTP Response: 
{
    "errorMessage": "State transition from Running to Ready failed for runtime. Error: State transition is not allowed",
    "errorType": "InvalidStateTransition"
}
Amazon.Lambda.RuntimeSupport.RuntimeApiClientException`1[Amazon.Lambda.RuntimeSupport.ErrorResponse]: Forbidden
Status: 403
Response: 
{
    "errorMessage": "State transition from Running to Ready failed for runtime. Error: State transition is not allowed",
    "errorType": "InvalidStateTransition"
}
   at Amazon.Lambda.RuntimeSupport.InternalRuntimeApiClient.NextAsync(String endpointUrl, CancellationToken cancellationToken) in /src/Repo/Libraries/src/Amazon.Lambda.RuntimeSupport/Client/InternalClientAdapted.cs:line 302
   at Amazon.Lambda.RuntimeSupport.RuntimeApiClient.RestoreNextInvocationAsync(CancellationToken cancellationToken) in /src/Repo/Libraries/src/Amazon.Lambda.RuntimeSupport/Client/RuntimeApiClient.cs:line 155
   at Amazon.Lambda.RuntimeSupport.Helpers.SnapstartHelperInitializeWithSnapstartIsolatedAsync.InitializeWithSnapstartAsync(IRuntimeApiClient client, Object restoreHooksRegistry) in /src/Repo/Libraries/src/Amazon.Lambda.RuntimeSupport/Helpers/SnapstartHelperInitializeWithSnapstartIsolatedAsync.cs:line 29
```

So essentially its running my custom code in RegisterBeforeSnapshot and then fails immediately after.

Im interested to know if anyone can suggest a path to approach solving this issue? if indeed it is something Im doing incorrectly...

**Comment by Beau-Gosse-dev** [comment]
We'll look into this and may request more info.

**Reply by Beau-Gosse-dev** [reply]
@tmitchel2 , can you provide details on how you're doing the following:
1. Registering the hooks with RegisterBeforeSnapshot and/or RegisterAfterRestore. Where are you making these calls? What do the calls look like?
2. How do you enable SnapStart for the function? 
3. Is this error getting thrown when publishing a new version of the Lambda? If not, when does it get thrown? 
4. Can we see the full stack trace?
5. Are you doing a class-library deployment inside a container? So your code doesn't have a main method?

**Reply by tmitchel2** [reply]
1.  This is my LambdaEntryPoint, 

```
using System;
using System.Diagnostics.CodeAnalysis;
using System.Text.Json;
using System.Text.Json.Serialization;
using System.Threading.Tasks;
using Amazon.Lambda.Serialization.SystemTextJson;
using Destash.Services;
using Microsoft.AspNetCore.Builder;
using Microsoft.Extensions.DependencyInjection;

namespace Destash.Lambda.AspNetCoreServer.APIGatewayProxyFunction
{
    public static class LambdaEntryPoint
    {
        [UnconditionalSuppressMessage("Trimming", "IL2026:Members annotated with 'RequiresUnreferencedCodeAttribute' require dynamic access otherwise can break functionality when trimming application code", Justification = "<Pending>")]
        public static async Task FunctionHandlerAsync()
        {
            var builder = WebApplication.CreateSlimBuilder();
            builder.Services.ConfigureHttpJsonOptions(options =>
            {
                options.SerializerOptions.TypeInfoResolverChain.Insert(0, DestashJsonContext.Default);
            });

            builder
                .Services
                .AddAWSLambdaHosting(LambdaEventSource.RestApi, hostingOptions =>
                {
                    hostingOptions.Serializer = new SourceGeneratorLambdaJsonSerializer<AspNetCoreServerJsonSerializer>(
                        serializerOptions =>
                        {
                            serializerOptions.PropertyNamingPolicy = JsonNamingPolicy.CamelCase;
                            serializerOptions.DefaultIgnoreCondition = JsonIgnoreCondition.WhenWritingNull;
                        });
                })
                .AddDestashWebServer(builder.Configuration);

            var app = builder.Build();
            app.UseDestash();
            app.MapDestashEndpoints();
            Amazon.Lambda.Core.SnapshotRestore.RegisterBeforeSnapshot(() => BeforeCheckpointAsync(app));
            Amazon.Lambda.Core.SnapshotRestore.RegisterAfterRestore(AfterCheckpoint);
            await app.RunAsync();
        }

        private static async ValueTask BeforeCheckpointAsync(WebApplication app)
        {
            Console.WriteLine($"BEFORE_CHECKPOINT_START TraceId: {TryGetTraceId()} Version: {TryGetFunctionVersion()} ProcessId: {Environment.ProcessId}");
            // UserCodeInit.PrepareAssembly(typeof(Program).Assembly);
            await app
                .Services
                .GetRequiredService<IDestashHostInitialisation>()
                .WaitUntilInitialisedAsync();
            Console.WriteLine($"BEFORE_CHECKPOINT_END TraceId: {TryGetTraceId()} Version: {TryGetFunctionVersion()} ProcessId: {Environment.ProcessId}");
        }

        private static async ValueTask AfterCheckpoint()
        {
            Console.WriteLine($"AFTER_CHECKPOINT_START TraceId: {TryGetTraceId()} Version: {TryGetFunctionVersion()} ProcessId: {Environment.ProcessId}");
            await ValueTask.CompletedTask;
            Console.WriteLine($"AFTER_CHECKPOINT_END TraceId: {TryGetTraceId()} Version: {TryGetFunctionVersion()} ProcessId: {Environment.ProcessId}");
        }

        private static string? TryGetTraceId()
        {
            return Environment.GetEnvironmentVariable("_X_AMZN_TRACE_ID");
        }

        private static string? TryGetFunctionVersion()
        {
            return Environment.GetEnvironmentVariable("AWS_LAMBDA_FUNCTION_VERSION");
        }
    }
}
```

2. I use CDK for deployment and enabling snapstart

```
const functionHandlerInner = new Function(
      this,
      "DestashAppServerHandlerFunction",
      {
        description: "APIGatewayProxyFunction handler",
        architecture: props.useArmArchitecture
          ? Architecture.ARM_64
          : Architecture.X86_64,
        runtime: Runtime.DOTNET_8,
        code: Code.fromAsset(path.join(__dirname, "../../../csharp"), {
          bundling: {
            image: Runtime.DOTNET_8.bundlingImage,
            workingDirectory: "/asset-input",
            user: "root",
            outputType: BundlingOutput.ARCHIVED,
            command: [
              "/bin/sh",
              "-c",
              " dotnet tool install -g Amazon.Lambda.Tools" +
                " && dotnet build" +
                " && dotnet lambda package --configuration Release --function-architecture arm64 --package-type zip --project-location ./Destash.Server --output-package /asset-output/function.zip",
            ],
          },
        }),
        handler:
          "Destash.Server::Destash.Lambda.AspNetCoreServer.APIGatewayProxyFunction.LambdaEntryPoint::FunctionHandlerAsync",
        timeout: Duration.seconds(30),
        memorySize: 5308, // 4 cores
        tracing: Tracing.ACTIVE,
        logGroup: this.apiLogGroup,
        environment: getEnvironmentVariables(props),
        // snapStart: SnapStartConf.ON_PUBLISHED_VERSIONS,
      }
    );
```

3. The error is not during deployment.  It gets thrown on all invocations to the newly deployed functions.
4. Im going to turn snapstart back on without any BeforeSnapshot or AfterRestore hooks and get some better logs for you.
5. It's a bit of a hybrid.  My code has a class library with this particular LambdaEntryPoint class and FunctionHandlerAsync method in it, but it does also have a runnable main method which can be used to invoke this and other functions which is useful for container deployments.

**Reply by bhallasaksham** [reply]
1. Thanks for sharing your code @tmitchel2, I'm curious if you built this using existing [blueprints](https://github.com/aws/aws-lambda-dotnet/tree/master/Blueprints/BlueprintDefinitions/vs2022)?
2. your cdk looks good, i doubt anything needs to change for SnapStart.
3. Ah, just to confirm, deploying this code and invoking with SnapStart disabled also doesn't work? None of the snapstart/hooks code should be executing if snapstart is disabled, so I'm curious what the error is when Snapstart is disabled. 
4. If you see the error only when SnapStart is enabled, it would be helpful to know whether its during publishing a version, or invoking the function. The BeforeSnapshot hooks should only be executing during version publish, and not during subsequent invokes.

**Reply by tmitchel2** [reply]
Here are the logs... These have snapstart turned on and it DOES have the Before and After hooks registered.  It looks as though the state transition error bubbles up causing the hosting to fail to start.

```
2024-11-20T19:25:04.745Z		info	BEFORE_CHECKPOINT_END TraceId: Root=1-673e3787-7ef554344a1976cc578db24b;Parent=9a631270085ebe05;Sampled=1;Lineage=1:dd7e3d1b:0 Version: 17 ProcessId: 10
```

```
2024-11-20T19:25:04.764Z		fail	Failed to invoke before snapshot hooks: HTTP Response: 
{
    "errorMessage": "State transition from Running to Ready failed for runtime. Error: State transition is not allowed",
    "errorType": "InvalidStateTransition"
}
Amazon.Lambda.RuntimeSupport.RuntimeApiClientException`1[Amazon.Lambda.RuntimeSupport.ErrorResponse]: Forbidden
Status: 403
Response: 
{
    "errorMessage": "State transition from Running to Ready failed for runtime. Error: State transition is not allowed",
    "errorType": "InvalidStateTransition"
}
   at Amazon.Lambda.RuntimeSupport.InternalRuntimeApiClient.NextAsync(String endpointUrl, CancellationToken cancellationToken) in /src/Repo/Libraries/src/Amazon.Lambda.RuntimeSupport/Client/InternalClientAdapted.cs:line 302
   at Amazon.Lambda.RuntimeSupport.RuntimeApiClient.RestoreNextInvocationAsync(CancellationToken cancellationToken) in /src/Repo/Libraries/src/Amazon.Lambda.RuntimeSupport/Client/RuntimeApiClient.cs:line 155
   at Amazon.Lambda.RuntimeSupport.Helpers.SnapstartHelperInitializeWithSnapstartIsolatedAsync.InitializeWithSnapstartAsync(IRuntimeApiClient client, Object restoreHooksRegistry) in /src/Repo/Libraries/src/Amazon.Lambda.RuntimeSupport/Helpers/SnapstartHelperInitializeWithSnapstartIsolatedAsync.cs:line 29
HTTP Response: 
{
    "errorMessage": "State transition from Running to Ready failed for runtime. Error: State transition is not allowed",
    "errorType": "InvalidStateTransition"
}
Amazon.Lambda.RuntimeSupport.RuntimeApiClientException`1[Amazon.Lambda.RuntimeSupport.ErrorResponse]: Forbidden
Status: 403
Response: 
{
    "errorMessage": "State transition from Running to Ready failed for runtime. Error: State transition is not allowed",
    "errorType": "InvalidStateTransition"
}
   at Amazon.Lambda.RuntimeSupport.InternalRuntimeApiClient.NextAsync(String endpointUrl, CancellationToken cancellationToken) in /src/Repo/Libraries/src/Amazon.Lambda.RuntimeSupport/Client/InternalClientAdapted.cs:line 302
   at Amazon.Lambda.RuntimeSupport.RuntimeApiClient.RestoreNextInvocationAsync(CancellationToken cancellationToken) in /src/Repo/Libraries/src/Amazon.Lambda.RuntimeSupport/Client/RuntimeApiClient.cs:line 155
   at Amazon.Lambda.RuntimeSupport.Helpers.SnapstartHelperInitializeWithSnapstartIsolatedAsync.InitializeWithSnapstartAsync(IRuntimeApiClient client, Object restoreHooksRegistry) in /src/Repo/Libraries/src/Amazon.Lambda.RuntimeSupport/Helpers/SnapstartHelperInitializeWithSnapstartIsolatedAsync.cs:line 29
```

```
2024-11-20T19:25:04.777Z
{
    "timestamp": "2024-11-20T19:25:04.7745461+00:00",
    "logLevel": "error",
    "message": "Hosting failed to start",
    "exception": "HTTP Response: \n\n{\"errorMessage\":\"State transition from Running to InitError failed for runtime. Error: State transition is not allowed\",\"errorType\":\"InvalidStateTransition\"}\n\n\nAmazon.Lambda.RuntimeSupport.RuntimeApiClientException`1[Amazon.Lambda.RuntimeSupport.ErrorResponse]: Forbidden\n\nStatus: 403\nResponse: \n{\"errorMessage\":\"State transition from Running to InitError failed for runtime. Error: State transition is not allowed\",\"errorType\":\"InvalidStateTransition\"}\n\n   at Amazon.Lambda.RuntimeSupport.InternalRuntimeApiClient.ErrorAsync(String lambda_Runtime_Function_Error_Type, String errorJson, String url, CancellationToken cancellationToken) in /src/Repo/Libraries/src/Amazon.Lambda.RuntimeSupport/Client/InternalClientAdapted.cs:line 192\n   at Amazon.Lambda.RuntimeSupport.Helpers.SnapstartHelperInitializeWithSnapstartIsolatedAsync.InitializeWithSnapstartAsync(IRuntimeApiClient client, Object restoreHooksRegistry) in /src/Repo/Libraries/src/Amazon.Lambda.RuntimeSupport/Helpers/SnapstartHelperInitializeWithSnapstartIsolatedAsync.cs:line 35\n   at Amazon.Lambda.RuntimeSupport.LambdaBootstrap.RunAsync(CancellationToken cancellationToken) in /src/Repo/Libraries/src/Amazon.Lambda.RuntimeSupport/Bootstrap/LambdaBootstrap.cs:line 180\n   at Microsoft.AspNetCore.Hosting.GenericWebHostService.StartAsync(CancellationToken cancellationToken)\n   at Microsoft.Extensions.Hosting.Internal.Host.<StartAsync>b__15_1(IHostedService service, CancellationToken token)\n   at Microsoft.Extensions.Hosting.Internal.Host.ForeachService[T](IEnumerable`1 services, CancellationToken token, Boolean concurrent, Boolean abortOnFirstException, List`1 exceptions, Func`3 operation)",
    "state": {
        "message": "Hosting failed to start",
        "{OriginalFormat}": "Hosting failed to start"
    },
    "eventId": 11,
    "category": "Microsoft.Extensions.Hosting.Internal.Host",
    "xrayTraceId": "1-673e3787-7ef554344a1976cc578db24b"
}
```

```
2024-11-20T19:25:04.785Z	a04dc601-c92d-4c59-ac0d-3a71bed9816e	fail	HTTP Response: 
{
    "errorMessage": "State transition from Running to InitError failed for runtime. Error: State transition is not allowed",
    "errorType": "InvalidStateTransition"
}
Amazon.Lambda.RuntimeSupport.RuntimeApiClientException`1[Amazon.Lambda.RuntimeSupport.ErrorResponse]: Forbidden
Status: 403
Response: 
{
    "errorMessage": "State transition from Running to InitError failed for runtime. Error: State transition is not allowed",
    "errorType": "InvalidStateTransition"
}
   at Amazon.Lambda.RuntimeSupport.InternalRuntimeApiClient.ErrorAsync(String lambda_Runtime_Function_Error_Type, String errorJson, String url, CancellationToken cancellationToken) in /src/Repo/Libraries/src/Amazon.Lambda.RuntimeSupport/Client/InternalClientAdapted.cs:line 192
   at Amazon.Lambda.RuntimeSupport.Helpers.SnapstartHelperInitializeWithSnapstartIsolatedAsync.InitializeWithSnapstartAsync(IRuntimeApiClient client, Object restoreHooksRegistry) in /src/Repo/Libraries/src/Amazon.Lambda.RuntimeSupport/Helpers/SnapstartHelperInitializeWithSnapstartIsolatedAsync.cs:line 35
   at Amazon.Lambda.RuntimeSupport.LambdaBootstrap.RunAsync(CancellationToken cancellationToken) in /src/Repo/Libraries/src/Amazon.Lambda.RuntimeSupport/Bootstrap/LambdaBootstrap.cs:line 180
   at Microsoft.AspNetCore.Hosting.GenericWebHostService.StartAsync(CancellationToken cancellationToken)
   at Microsoft.Extensions.Hosting.Internal.Host.<StartAsync>b__15_1(IHostedService service, CancellationToken token)
   at Microsoft.Extensions.Hosting.Internal.Host.ForeachService[T](IEnumerable`1 services, CancellationToken token, Boolean concurrent, Boolean abortOnFirstException, List`1 exceptions, Func`3 operation)
   at Microsoft.Extensions.Hosting.Internal.Host.StartAsync(CancellationToken cancellationToken)
   at Microsoft.Extensions.Hosting.HostingAbstractionsHostExtensions.RunAsync(IHost host, CancellationToken token)
   at Microsoft.Extensions.Hosting.HostingAbstractionsHostExtensions.RunAsync(IHost host, CancellationToken token)
   at Destash.Lambda.AspNetCoreServer.APIGatewayProxyFunction.LambdaEntryPoint.FunctionHandlerAsync() in /asset-input/Destash.Server/Lambda.AspNetCoreServer.APIGatewayProxyFunction/LambdaEntryPoint.cs:line 42
   at lambda_method1(Closure, Stream, ILambdaContext, Stream)
   at Amazon.Lambda.RuntimeSupport.HandlerWrapper.<>c__DisplayClass8_0.<GetHandlerWrapper>b__0(InvocationRequest invocation) in /src/Repo/Libraries/src/Amazon.Lambda.RuntimeSupport/Bootstrap/HandlerWrapper.cs:line 54
   at Amazon.Lambda.RuntimeSupport.LambdaBootstrap.InvokeOnceAsync(CancellationToken cancellationToken) in /src/Repo/Libraries/src/Amazon.Lambda.RuntimeSupport/Bootstrap/LambdaBootstrap.cs:line 237
```

```
END RequestId: a04dc601-c92d-4c59-ac0d-3a71bed9816e
```

**Reply by tmitchel2** [reply]
To answer your follow up questions
1) I know of the blueprints, it mainly follows the NativeAOT & Custom Runtime approaches but it's evolved through time.
2) Great
3) Disabling snapstart with all the logic the same works correctly.  The hooks don't get invoked as you would expect and they are optional for how my app works so all good there from my point of view.
4) This is interesting, I presumed the snapshot was created on the first invoke.  How do I go about finding logs relevant to when its being published to see if it ran my code correctly?  Any particular log message that I should see?

**Reply by bhallasaksham** [reply]
Snapshot is created when you publish a version. You should be able to see the `INIT` logs that invoke the before snapshot hooks in CloudWatch. You can always enable debug logging by setting `LAMBDA_RUNTIMESUPPORT_DEBUG` to true in your environment variables.

**Reply by tmitchel2** [reply]
Ok i see only the following two lines in the init log streams. I presume that is wrong and i should see my startup logs?  Im going to enable LAMBDA_RUNTIMESUPPORT_DEBUG and see if that turns up anything...

```
2024-11-20T19:21:48.633Z | INIT_START Runtime Version: dotnet:8.v29 Runtime Version ARN: arn:aws:lambda:eu-west-1::runtime:blahblahblah
2024-11-20T19:21:49.099Z | INIT_REPORT Init Duration: 466.82 ms
```

**Reply by bhallasaksham** [reply]
Few more questions - 
- How are you enabling SnapStart? Are you manually setting any environment variables?
- Are you using NativeAOT?

**Reply by bhallasaksham** [reply]
Seems like the `BeforeSnapshot` hooks are running during the `RESTORE` phase, and not during `INIT` phase?

**Reply by tmitchel2** [reply]
I turned bootstrap logging on and it's made me understand the process a lot more and given me a direction.

I think the confusion arises because I'm using minimal API style / nativeaot compatible approach (but obviously not turned on for this particular deployment).  The documentation for snap start is thin for minimal API from what I can see and I assumed that the hooks should be wired up before the RunAsync, but for this approach that is wrong, it should be in the constructor.  Seeing the snapshot hooks getting fired also threw me off considering it was restored already.

Anyway I will play around with it again and refer back once I have it working.

**Reply by tmitchel2** [reply, marked as answer]
Ok so I've resolved the issue, the LAMBDA_RUNTIMESUPPORT_DEBUG was very helpful so I appreciate you pointing that out to me.

My issue arose because I was using a Executable Assembly style but my cdk handler was configured for Class Library approach.  Interestingly that still worked with snapstart disabled and thats what caused the confusion.  When SnapStart was enabled, it would take over and it would search for a constructor that wasnt there in the LambdaEntryPoint, and thus wouldnt register the hooks and wouldnt run any of my code.

TLDR I made some other changes, but the main fix for me was to update my handler to the execution assembly style:
```
handler: "Destash.Server"
```
